request-manager/graph/grapher.go

```diff
@@ -217,25 +236,27 @@ func buildSeqGraph(seqSpec *spec.Sequence, idgen id.Generator) (seqGraph *Graph,
 		}
 	}
 
-	// Nodes we've yet to add
+	// Nodes we've yet to add, represented as single-node graphs
 	nodesToAdd := map[string]*Graph{}
 	for k, v := range nodes {
 		nodesToAdd[k] = v
 	}
 	nodesAdded := map[string]bool{}
 
 	// Build graph by adding nodes, starting from the source node, and then
-	// adding all adjacent nodes to the source node, and so on.
-	// We cannot add nodes in any order because we do not know the reverse dependencies
+	// adding all adjacent nodes to the source node, and so on
+	// We cannot add nodes in any order because we do not know the reverse
+	// dependencies
```

Comment:
Good comment again. We can make it a little better. The most important thing for future engineers to know here: the nodes in the request spec are _not_ in topological order, or any order, i.e. spec order != graph order. In spec, graph order is implied by the `deps` field. So here in code, we continuously loop through the (potentially random) spec order, creating nodes if and when all dependent nodes have been created. Usually , spec order is close to topological order, but it's not a rule.